brion/circuitConfig.cpp

```diff
@@ -19,208 +19,71 @@
-                auto& vJValue = vJ.second;
-                auto startPos = vJValue.find(vIKey);
-
-                if (startPos != std::string::npos)
-                {
-                    vJValue.replace(startPos, vIKey.length(), vIValue);
-                    anyChange = true;
-                }
-            }
-        }
-
-        variables = variablesCopy;
-        iteration++;
-    }
-
-    if (iteration == maxIterations)
-        throw std::runtime_error(
-            "Reached maximum allowed iterations in variable expansion, "
-            "possibly infinite recursion.");
-
-    return variables;
-}
-
-brion_nlohmann::json _expandVariables(
-    const brion_nlohmann::json& json,
-    const std::map<std::string, std::string>& variables)
+namespace brion
 {
-    auto jsonFlat = json.flatten();
-
-    // Expand variables in whole json
-    for (auto it = jsonFlat.begin(); it != jsonFlat.end(); ++it)
-    {
-        if (!it.value().is_string())
-            continue;
-
-        auto valueStr = it.value().get<std::string>();
-        auto& value = it.value();
-
-        for (auto& var : variables)
-        {
-            auto& varName = var.first;
-            auto& varValue = var.second;
-            auto startPos = valueStr.find(varName);
-
-            if (startPos != std::string::npos)
-            {
-                valueStr.replace(startPos, varName.length(), varValue);
-                value = valueStr;
-            }
-        }
-    }
-
-    return jsonFlat.unflatten();
-}
-
-brion_nlohmann::json _parseCircuitJson(const std::string& jsonStr)
+namespace
 {
-    const auto json = brion_nlohmann::json::parse(jsonStr);
-
-    auto variables = _readVariables(json);
-    variables = _resolveVariables(variables);
-    return _expandVariables(json, variables);
-}
-
-std::map<std::string, std::string> _fillComponents(
-    const brion_nlohmann::json& json, const boost::filesystem::path& basePath)
+std::map<std::string, std::string> _fillComponents(const nlohmann::json& json,
+                                                   const PathResolver& resolver)
 {
-    const auto comps = json["components_dir"];
+    const auto comps = json.at("components");
     std::map<std::string, std::string> output;
 
     for (auto it = comps.begin(); it != comps.end(); ++it)
-        output[it.key()] = toAbsolute(it.value(), basePath);
+        output[it.key()] = resolver.toAbsolute(it.value());
 
     return output;
 }
 
 std::vector<brion::CircuitConfig::SubnetworkFiles> _fillSubnetwork(
-    const brion_nlohmann::json& json, const std::string& networkType,
+    const nlohmann::json& json, const std::string& networkType,
     const std::string& elementName, const std::string& typeName,
-    const boost::filesystem::path& basePath)
+    const PathResolver& resolver)
 {
     std::vector<brion::CircuitConfig::SubnetworkFiles> output;
 
-    const auto nodes = json["networks"][networkType];
+    const auto nodes = json.at("networks").at(networkType);
```

Comment:
I replaced the previous code with .at because it's already better than crashing when a bad input file is given. Replacing the .at with find + something else wouldn't be better. This function is called from a constructor and you can only make the constructor fail by using an exception, so that something else can only be to throw eventually.
The exception will be propagated out of the brain::Circuit constructor and it's up to the user to catch it or ignore it. I think this is reasonable because the library can't do anything about bad inputs. In the Python wrappings this is enough to not crash the interpreter and having a fairly meaningful error message. I can translate the exception into another type, but I prefer not doing so given that we typically do nothing for these type of fatal errors (other that trying not to crash).